src/qibo/models/error_mitigation.py

```diff
@@ -592,8 +632,340 @@ def apply_randomized_readout_mitigation(
+        noise_model (qibo.models.noise.Noise, optional): the noise model to be applied. Defaults to ``None``.
+        nshots (int, optional): the number of shots for the circuit execution. Defaults to :math:`10000`.
+        readout (dict, optional): a dictionary that may contain the following keys:
+
+            *    ncircuits: int, specifies the number of random circuits to use for the randomized method of readout error mitigation.
+            *    response_matrix: numpy.ndarray, used for applying a pre-computed response matrix for readout error mitigation.
+            *    ibu_iters: int, specifies the number of iterations for the iterative Bayesian unfolding method of readout error mitigation. If provided, the corresponding readout error mitigation method is used. Defaults to {}.
+
+        qubit_map (list, optional): the qubit map. If None, a list of range of circuit's qubits is used. Defaults to ``None``.
+        backend (qibo.backends.abstract.Backend, optional): the backend to be used in the execution.
+            If None, it uses the global backend. Defaults to ``None``.
+
+    Returns:
+        float: The mitigated expectation value of the observable.
+    """
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    if "ncircuits" in readout:
+        circuit_result, circuit_result_cal = apply_randomized_readout_mitigation(
+            circuit, noise_model, nshots, readout["ncircuits"], backend
+        )
+    else:
+        circuit_result = _circuit_conf(
+            circuit, qubit_map, noise_model, nshots, backend=backend
+        )
+
+    if "response_matrix" in readout:
+        circuit_result = apply_resp_mat_readout_mitigation(
+            circuit_result,
+            readout["response_matrix"],
+            readout.get("ibu_iters", None),
+        )
+
+    exp_val = circuit_result.expectation_from_samples(observable)
+
+    if "ncircuits" in readout:
+        exp_val /= circuit_result_cal.expectation_from_samples(observable)
+
+    return exp_val
+
+
+def sample_clifford_training_circuit(
+    circuit,
+    backend=None,
+):
+    """Samples a training circuit for CDR by susbtituting some of the non-Clifford gates.
+
+    Args:
+        circuit (:class:`qibo.models.Circuit`): circuit to sample from.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): backend to be used
+            in the execution. If ``None``, it uses :class:`qibo.backends.GlobalBackend`.
+            Defaults to ``None``.
+
+    Returns:
+        :class:`qibo.models.Circuit`: The sampled circuit.
+    """
+    from qibo.quantum_info import (  # pylint: disable=import-outside-toplevel
+        random_clifford,
+    )
+
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    non_clifford_gates = [
+        (i, gate)
+        for i, gate in enumerate(circuit.queue)
+        if not gate.clifford and not isinstance(gate, gates.M)
+    ]
```

Comment:
I don't think the gates of this tuples are used anywhere in the following
```suggestion
    non_clifford_gates_indices = [
        i
        for i, gate in enumerate(circuit.queue)
        if not gate.clifford and not isinstance(gate, gates.M)
    ]
```